### PR TITLE
Fix automorphisms for Hypercube and initialization bug in RbmSpinSymm

### PR DESCRIPTION
**Pull request by gcarleo**
Also change python names for machine and add test for log derivatives

### CODE REVIEW

File: NetKet/Graph/hypercube.hpp [hunk 1]
```diff
@@ -218,30 +218,43 @@ class Hypercube : public AbstractGraph {
...
-        do {
-          auto &i = coord[dim];
-          auto const old_i = i;
-          i = (i + rotate_by) % L;
-          permutation.push_back(Coord2Site(coord, L));
-          i = old_i;
-        } while (next_variation(coord.rbegin(), coord.rend(), L - 1));
-
-        symm_table.emplace_back(std::move(permutation));
+    if (!pbc) {
```
Comment (twesterhout) [inline NetKet/Graph/hypercube.hpp]: Indeed, I forgot this case!

File: NetKet/Graph/hypercube.hpp [hunk 2]
```diff
@@ -218,30 +218,43 @@ class Hypercube : public AbstractGraph {
   static std::vector<std::vector<int>> BuildSymmTable(int const L,
                                                       int const n_dim, bool pbc,
                                                       int const n_sites) {
-    if (!pbc) return {};
-    assert(L >= 2 && n_dim >= 1 &&
-           "Bug! This should be the case by construction.");
-    std::vector<std::vector<int>> symm_table;
```
Comment (twesterhout) [inline NetKet/Graph/hypercube.hpp]: But what's wrong with the rest of the implementation?

Comment (gcarleo) [reply NetKet/Graph/hypercube.hpp]: I am not sure, but your implementation gives different permtables for square or cubic lattices. I think you are not considering all the possible translation vectors, but I am not sure

**Comment by twesterhout** [comment]
@gcarleo ah, I see what you mean... I also understand where the mistake is, but let's just merge this, because the whole reason for an alternative solution was simplicity and I'm not sure it'll remain just as simple if I fix the bug. Good job noticing this!

As for the Machine part, I'm not sure I like the exposed interface, but I'll complain about that later :)

**Pull request merged**